ErrorHandling/src/backend/useBackendService.ts

```diff
@@ -0,0 +1,129 @@
+import crashlytics from '@react-native-firebase/crashlytics';
+import axios, {AxiosError} from 'axios';
+import {Alert} from 'react-native';
+import {
+  EnsuredQueryKey,
+  QueryClient,
+  QueryKey,
+  useMutation,
+  UseMutationOptions,
+  UseMutationResult,
+  useQuery,
+  useQueryClient,
+  UseQueryOptions,
+  UseQueryResult,
+} from 'react-query';
+
+import {HttpApiClient} from './HttpApiClient';
+
+// 画面の入力値によってbaseURLを変えるので、baseURLを設定しないaxiosのインスタンスでHTTP APIクライアントを作成
+const httpApiClient = new HttpApiClient(axios.create());
+
+export type ErrorResponse = {
+  code: string;
+  message: string;
+};
+
+export const useBackendQuery = <TData = unknown, TQueryKey extends QueryKey = QueryKey>(
+  queryKey: TQueryKey,
+  pathFn: (queryKey: EnsuredQueryKey<TQueryKey>) => string,
+  options?: UseQueryOptions<TData, AxiosError<ErrorResponse>, TData, TQueryKey>,
+): UseQueryResult<TData, AxiosError<ErrorResponse>> => {
+  const queryClient = useQueryClient();
+  return useQuery<TData, AxiosError<ErrorResponse>, TData, TQueryKey>(
+    queryKey,
+    async (context) => {
+      const CancelToken = axios.CancelToken;
+      const source = CancelToken.source();
+      const promise = httpApiClient.get<TData>(pathFn(context.queryKey), source.token);
+      // @ts-ignore
+      promise.cancel = () => {
+        source.cancel('Query was cancelled by React Query');
+      };
+      const timeoutObj = timeout(queryClient, queryKey);
+      try {
+        return (await promise).data;
+      } finally {
+        clearTimeout(timeoutObj);
+      }
+    },
+    {
+      onError: (err) => _onError(err),
+      retry: _retry,
+      ...options,
+    },
+  );
+};
+
+export const useBackendMutation = <TData = unknown, TVariables = void, TContext = unknown>(
+  path: string,
+  options?: UseMutationOptions<TData, AxiosError<ErrorResponse>, TVariables, TContext>,
+): UseMutationResult<TData, AxiosError<ErrorResponse>, TVariables, TContext> => {
+  return useMutation<TData, AxiosError<ErrorResponse>, TVariables, TContext>(
+    async (variables) => {
+      const response = await httpApiClient.post<TData>(path, variables);
+      return response.data;
+    },
+    {
+      onError: (err) => _onError(err),
```

Comment:
このI/Fだと、自分たちでエラーハンドリングするときに必ず全ケースのハンドリングをしなきゃいけない気がする。チェーンみたいなことができたりしないのかな？

Reply:
個別にハンドリングする場合は、useBackendQuery、useBackendMutationから返却される`error`を見てハンドリングするイメージ

Postの時の`ErrorInHttpApi`のハンドリング方法を間違って実装してたので、修正しましたー
e5adba2581b31322ce5e424a7a47232d16da2bca

Reply:
関数そのままぽいってしたほうが良い気がする。なんなら、`_onError`じゃなくて`onError`にしてしまってもよいのでは

```suggestion
      onError,
```

Reply:
ほいっ
ca265c2199b1da59cc5005863ae6983394132824

Reply:
おしいw

Reply:
あっw

2f53a34f5d83ded21ccf34af316801e3d27b567e